### PR TITLE
Add noetic to CI, improvements to type safety checking

### PR DESCRIPTION
**Pull request by valgur**
I added noetic to the CI config and turned on `-Werror` for CI. The newer Ubuntu release in noetic also includes a newer gcc with improved warnings. Pretty much all of the warnings (except for #5) were related to the `isCompatible()` and `inBounds()` checks in `messages.cpp`:
* warning: implicit conversion from 'int' to 'float' changes value from 2147483647 to 2147483648.
* warning: comparison between signed and unsigned integer expressions.

Since that code does some tricky implicit casting and the warnings seemed relevant, I did not want to just dismiss them. However, I found the existing rules making use of SFINAE a bit hard to follow, so I created a simple test suite for `isCompatible()` that checks all arithmetic type pairs and rewrote that check from scratch.

`isCompatible()` now effectively checks that the conversion is guaranteed to safe:
* Integer types can safely fit within the other.
* Integer to floating point type conversion can be done without losing precision - that is, the integer can fit inside the mantissa.
* Floating point to integer is not permitted.
* Floating point to floating point is only allowed to the same or bigger type.

I also made `isCompatible()` and `inBounds()` constexpr and got rid of `isCompatible(const T &val)` and `m->type() != message_type<U>::value`. The latter is redundant to `isCompatible()` and thanks to `isCompatible()` being constexpr, most type conversion safety checks can be done at compile-time.

### CODE REVIEW

**Comment by StefanFabian** [comment]
Wow, thank you for this refactoring!
Since it contains quite a bit of logic changes, it will take a few days to properly review the changes.
I will add some small comments now and review the rest later this week.

File: ros_babel_fish/src/message.cpp [hunk 1]
```diff
@@ -30,104 +30,101 @@ Message::~Message() = default;
...
+  }
+  m->as<ValueMessage<U>>().setValue( static_cast<U>(value) );
 }
 
 template<typename T>
 void assignToValueMessage( Message *m, const T &value )
 {
   using namespace message_type_traits;
-  switch ( m->type())
+  switch ( m->type() )
```
Comment (StefanFabian) [inline ros_babel_fish/src/message.cpp]: The code style doesn't use a space between parentheses.
If you are using CLion, you can use the attached zipped code style.
[ROS.zip](https://github.com/StefanFabian/ros_babel_fish/files/7152586/ROS.zip)

File: ros_babel_fish/test/message.cpp [hunk 2]
```diff
@@ -691,6 +691,142 @@ TEST( MessageTest, arrayMessage )
...
+  ASSERT_EQ( (isCompatible<int32_t, uint8_t>()), false );
+  ASSERT_EQ( (isCompatible<int32_t, uint16_t>()), false );
+  ASSERT_EQ( (isCompatible<int32_t, uint32_t>()), false );
+  ASSERT_EQ( (isCompatible<int32_t, uint64_t>()), true );
+  ASSERT_EQ( (isCompatible<int32_t, int8_t>()), false );
+  ASSERT_EQ( (isCompatible<int32_t, int16_t>()), false );
+  ASSERT_EQ( (isCompatible<int32_t, int32_t>()), true );
+  ASSERT_EQ( (isCompatible<int32_t, int64_t>()), true );
+  ASSERT_EQ( (isCompatible<int32_t, float>()), false );
+  ASSERT_EQ( (isCompatible<int32_t, double>()), true );
```
Comment (StefanFabian) [inline ros_babel_fish/test/message.cpp]: I'm not sure if I agree that an assignment of (unsigned) integer types larger than 2 bytes to floating-point types shouldn't be possible (without warning).
The introduced error for the values that a float can not represent exactly is negligible in most cases IMO and if your message member type is floating-point, you have to expect some floating-point error.

File: ros_babel_fish/src/message.cpp (hunk 1 above)
Comment (valgur) [reply ros_babel_fish/src/message.cpp]: Thanks, I was indeed wondering how these should actually be formatted. I would highly recommend adding a `.clang-format` to the repo, which takes out the guesswork, but I understand that the current style is not well supported by it.

File: ros_babel_fish/test/message.cpp (hunk 2 above)
Comment (valgur) [reply ros_babel_fish/test/message.cpp]: On second thought, yes, I agree. That was too stringent and warning for conversions to floating point is an overkill. If the user asks for a float then that's most likely the precision they are happy with. I updated the code and tests.

File: ros_babel_fish/src/message.cpp (hunk 1 above)
Comment (StefanFabian) [reply ros_babel_fish/src/message.cpp]: Yes, I was thinking about adding one but I would have to check what is supported. I think in the past it wasn't really the same and it was a bit cumbersome to get CLion to use it.

Comment (valgur) [reply ros_babel_fish/src/message.cpp]: Clang-format support has been very seamless in CLion for a while now and it apparently supports exporting `.clang-format` directly from its formatting config as well. I also experimented with https://github.com/valgur/clang-guess-format but it fails due to clang-format missing an option to avoid adding a space between adjacent parentheses like CLion does by default.

File: ros_babel_fish/test/message.cpp [hunk 3]
```diff
@@ -691,6 +691,140 @@ TEST( MessageTest, arrayMessage )
...
+  ASSERT_EQ((isCompatible<uint64_t, int8_t>()), false );
+  ASSERT_EQ((isCompatible<uint64_t, int16_t>()), false );
+  ASSERT_EQ((isCompatible<uint64_t, int32_t>()), false );
+  ASSERT_EQ((isCompatible<uint64_t, int64_t>()), false );
+  ASSERT_EQ((isCompatible<uint64_t, float>()), true );
+  ASSERT_EQ((isCompatible<uint64_t, double>()), true );
+  ASSERT_EQ((isCompatible<int8_t, uint8_t>()), false );
+  ASSERT_EQ((isCompatible<int8_t, uint16_t>()), true );
+  ASSERT_EQ((isCompatible<int8_t, uint32_t>()), true );
+  ASSERT_EQ((isCompatible<int8_t, uint64_t>()), true );
```
Comment (StefanFabian) [inline ros_babel_fish/test/message.cpp]: This doesn't make sense to me. A signed integer is never compatible with an unsigned integer as the target.

File: ros_babel_fish/src/message.cpp [hunk 4]
```diff
@@ -30,96 +30,95 @@ Message::~Message() = default;
...
+typename std::enable_if<std::is_floating_point<T>::value, bool>::type
+constexpr inBounds( const T &val )
 {
-  if ( std::is_same<T, U>::value ) return true;
-  if ( std::is_floating_point<T>::value ) return false;
-  return inBounds<T, U>( val );
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wimplicit-int-float-conversion"
+  return std::numeric_limits<U>::min() <= val && val <= std::numeric_limits<U>::max();
+#pragma GCC diagnostic pop
```
Comment (StefanFabian) [inline ros_babel_fish/src/message.cpp]: Personally, I would prefer to make the conversion explicit where necessary (when U is not a floating point) and ensure that the conversion will be to a large integral type such as int64_t.

**Review by StefanFabian — CHANGES_REQUESTED** [review]

**Comment by StefanFabian** [comment]
Thanks for your patience. I have changed the test cases I marked, can you check that your code passes the tests?
The old code does pass now with some modifications of the old code where I wasn't happy with the behavior.

File: ros_babel_fish/test/message.cpp (hunk 3 above)
Comment (valgur) [reply ros_babel_fish/test/message.cpp]: Uh, yes, don't know how I missed that.

File: ros_babel_fish/src/message.cpp (hunk 4 above)
Comment (valgur) [reply ros_babel_fish/src/message.cpp]: Indeed, I agree. Replaced it with
```cpp
return static_cast<T>(std::numeric_limits<U>::min()) <= val && val <= static_cast<T>(std::numeric_limits<U>::max());
```

**Comment by StefanFabian** [comment]
Thanks, looks good to me! I'll probably move the `isCompatible` and probably the rest of the boundary checking to a header file to avoid code duplication and merge it tomorrow.

**Comment by StefanFabian** [comment]
I always forget that you should never name concrete dates :)
Waiting for the CI to approve, then I'll merge.
Do you have any other PRs in the pipeline I should wait for before I make a new release?

**Comment by valgur** [comment]
No problem, I'm not in a rush anywhere. :)
These are all the changes I had in mind. Don't have anything more coming up.
